File: 0054-spiral-matrix/0054-spiral-matrix.go
func spiralOrder(matrix [][]int) []int {
    var list = make([]int,(len(matrix)*len(matrix[0])))
    var index int = 0
    
    top:=0
    left:=0
    right:=len(matrix[0])-1
    bottom:=len(matrix)-1
 
    for top<=bottom && left<=right{
        for i:=left;i<=right;i++{
            list[index] = matrix[top][i]
            index++
        }
        top++
        
        if(!(top<=bottom && left<=right)){
            return list
        }
        for i:=top;i<=bottom;i++{
            list[index] = matrix[i][right]
            index++
        }
        right--
        
        if(!(top<=bottom && left<=right)){
            return list
        }
        for i:=right;i>=left;i--{
            list[index] = matrix[bottom][i]
            index++
        }
        bottom--
        
        if(!(top<=bottom && left<=right)){
            return list
        }
        for i:=bottom;i>=top;i--{
            list[index] = matrix[i][left]
            index++
        }
        left++
    }
    return list
}

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> list = new ArrayList<>();
        int top=0,left=0,right=matrix[0].length-1,bottom=matrix.length-1;
        
        while(top<=bottom && left<=right){
            for(int i=left;i<=right;i++){
                list.add(matrix[top][i]);
            }
            top++;
            if(!(top<=bottom && left<=right))
                return list;
            
            for(int i=top;i<=bottom;i++){
                list.add(matrix[i][right]);
            }
            right--;
            if(!(top<=bottom && left<=right))
                return list;
            
            for(int i=right;i>=left;i--){
                list.add(matrix[bottom][i]);
            }
            bottom--;
            if(!(top<=bottom && left<=right))
                return list;
            
            for(int i=bottom;i>=top;i--){
                list.add(matrix[i][left]);
            }
            left++;
            
        }
        return list;
    }
}